In [1]:
import numpy as np
import time

from load_data import unpickle
from sklearn.model_selection import KFold
from sklearn.neighbors import KNeighborsClassifier

In [2]:
def get_accuracy(predictions, actual):
    difference = predictions - actual
    number_predictions = predictions.shape[0]
    number_incorrect = np.count_nonzero(difference)
    number_correct = number_predictions - number_incorrect
    accuracy = number_correct / number_predictions
    return accuracy

In [3]:
data = unpickle("data/data_batch_1")

In [4]:
X, y = data[b'data'], np.array(data[b'labels'])

In [5]:
start_time = time.time()

classifier = KNeighborsClassifier(n_neighbors=5)
accuracies = []

kf = KFold(n_splits=5, shuffle=True)
for train_index, test_index in kf.split(X, y):
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y[train_index], y[test_index]

    classifier.fit(X_train, y_train)
    y_predict = classifier.predict(X_test)
    accuracy = get_accuracy(y_predict, y_test)
    accuracies.append(accuracy)
    
end_time = time.time()

In [6]:
elapsed_time = end_time - start_time
elapsed_time = round(elapsed_time, 3)  
print("Elapsed time:", elapsed_time)

mean_accuracy = np.array(accuracies).mean()
mean_accuracy = round(mean_accuracy, 3)
print("Mean accuracy:", mean_accuracy)

Elapsed time: 515.352
Mean accuracy: 0.28


Using kNN with 5 neighbors and 5-fold cross-validation with shuffling on 10,000 images took approximately 8.5 minutes and resulted in a mean accuracy of 0.28